From cf0133a24574bf1c915423bfb8b8764b8d3c061a Mon Sep 17 00:00:00 2001
From: jinghua <xigu@marvell.com>
Date: Wed, 9 Nov 2016 16:56:06 +0800
Subject: [PATCH 1134/2241] xhci: a3700: update XHCI driver for USB OTG support

- Armada3700 has got both USB Host and Device controller,
  and a dedicate USB_ID interrupt as a indication of USB
  mode switch between Host and Device.
- This patch updates generic XHCI driver to support USB
  OTG framework, with newly introduced compatible string
  "marvell,armada-3700-xhci-otg", it will be working in
  USB OTG mode, register itself to OTG PHY driver
  in probe routine, and only start to work until OTG PHY
  driver decides to move to USB Host mode by handling USB_ID
  interrupt, and call usb_add_hcd from it.
- Armada3700 could still work in USB Host only mode as usual
  with regular generic XHCI driver compatible string "generic-xhci".
- This update is for Armada3700 USB OTG mode specially, and
  might not be able to mainline.

Change-Id: Iff3fc79218c5fa1d2780fd8bef775a7afc28a347
Signed-off-by: jinghua <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33631
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
---
 Documentation/devicetree/bindings/usb/usb-xhci.txt |  4 +-
 drivers/usb/host/xhci-plat.c                       | 48 +++++++++++++++++-----
 2 files changed, 40 insertions(+), 12 deletions(-)

diff --git a/Documentation/devicetree/bindings/usb/usb-xhci.txt b/Documentation/devicetree/bindings/usb/usb-xhci.txt
index 86f67f0..175151a 100644
--- a/Documentation/devicetree/bindings/usb/usb-xhci.txt
+++ b/Documentation/devicetree/bindings/usb/usb-xhci.txt
@@ -3,8 +3,8 @@ USB xHCI controllers
 Required properties:
   - compatible: should be one of "generic-xhci",
     "marvell,armada-375-xhci", "marvell,armada-380-xhci",
-    "renesas,xhci-r8a7790", "renesas,xhci-r8a7791" (deprecated:
-    "xhci-platform").
+    "renesas,xhci-r8a7790", "renesas,xhci-r8a7791",
+    "marvell,armada-3700-xhci-otg" (deprecated: "xhci-platform").
   - reg: should contain address and length of the standard XHCI
     register set for the device.
   - interrupts: one XHCI interrupt should be described here.
diff --git a/drivers/usb/host/xhci-plat.c b/drivers/usb/host/xhci-plat.c
index d71e32a..aa4f6c8 100644
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -24,6 +24,7 @@
 #include "xhci.h"
 #include "xhci-mvebu.h"
 #include "xhci-rcar.h"
+#include <linux/usb/otg.h>
 
 static struct hc_driver __read_mostly xhci_plat_hc_driver;
 
@@ -177,14 +178,34 @@ static int xhci_plat_probe(struct platform_device *pdev)
 			goto put_usb3_hcd;
 	}
 
-	ret = usb_add_hcd(hcd, irq, IRQF_SHARED);
-	if (ret)
-		goto disable_usb_phy;
-
-	ret = usb_add_hcd(xhci->shared_hcd, irq, IRQF_SHARED);
-	if (ret)
-		goto dealloc_usb2_hcd;
+	if (of_device_is_compatible(pdev->dev.of_node,
+				    "marvell,armada-3700-xhci-otg")) {
+		/* If Armada3700 needs to enable OTG support, register XHCI
+		 * driver to OTG PHY, and wait for it to call usb_add_hcd
+		 * at the right time (start working in USB Host mode).
+		 */
+
+		if (hcd->usb_phy == NULL) {
+			dev_err(&pdev->dev, "unable to find OTG PHY\n");
+			goto disable_usb_phy;
+		}
+
+		hcd->irq = irq;
+
+		ret = otg_set_host(hcd->usb_phy->otg, &hcd->self);
+		if (ret) {
+			dev_err(&pdev->dev, "unable to register with OTG PHY\n");
+			goto disable_usb_phy;
+		}
+	} else {
+		ret = usb_add_hcd(hcd, irq, IRQF_SHARED);
+		if (ret)
+			goto disable_usb_phy;
 
+		ret = usb_add_hcd(xhci->shared_hcd, irq, IRQF_SHARED);
+		if (ret)
+			goto dealloc_usb2_hcd;
+	}
 	return 0;
 
 
@@ -213,10 +234,16 @@ static int xhci_plat_remove(struct platform_device *dev)
 	struct xhci_hcd	*xhci = hcd_to_xhci(hcd);
 	struct clk *clk = xhci->clk;
 
-	usb_remove_hcd(xhci->shared_hcd);
-	usb_phy_shutdown(hcd->usb_phy);
+	if (of_device_is_compatible(dev->dev.of_node,
+				    "marvell,armada-3700-xhci-otg")) {
+		otg_set_host(hcd->usb_phy->otg, NULL);
+	} else {
+		usb_remove_hcd(xhci->shared_hcd);
+		usb_phy_shutdown(hcd->usb_phy);
+
+		usb_remove_hcd(hcd);
+	}
 
-	usb_remove_hcd(hcd);
 	usb_put_hcd(xhci->shared_hcd);
 
 	if (!IS_ERR(clk))
@@ -272,6 +299,7 @@ static const struct of_device_id usb_xhci_of_match[] = {
 	{ .compatible = "marvell,armada-380-xhci"},
 	{ .compatible = "renesas,xhci-r8a7790"},
 	{ .compatible = "renesas,xhci-r8a7791"},
+	{ .compatible = "marvell,armada-3700-xhci-otg"},
 	{ },
 };
 MODULE_DEVICE_TABLE(of, usb_xhci_of_match);
-- 
2.7.4

